### DISCUSSION TITLE
Model slanted grain growth with GBAnisotropy using an inclination angle

### DISCUSSION THREAD

**Discussion by Hejie-Guo** [idaholab/moose#26500] [Q&A Modules: Phase field]
Hello, moose community, i am a freshman in grain growth and moose modeling. 

I am trying to model slanted temperature-dependent grain growth behavior from experimental obervations and first following grain growth model in the work of "Phase-field modeling of grain evolutions in additive manufacturing from nucleation, growth, to coarsening" , the anisotropic grain growth behavior is similar as the following figures in the article:
![image](https://github.com/idaholab/moose/assets/61571563/0b8d960b-de86-4ce5-aa8b-44f8fc6bcecb)

I simply modify some moose examples and modeling the temperature-dependent behavior, but did not quite understand how to model the anisotropic grain growth. Though i have test the examples in the dir ../GBAnisotropy/, I am still confused that how to model the slanted shape. Here is the result i can get using moose,  i just want to anisotropic modeling the shape as the black line.
![image](https://github.com/idaholab/moose/assets/61571563/6554db46-f538-4b4f-a8cd-4bdfcffb3d35)

The grain growth model in the article may be explicitly modeled by moose and do not need additional coding work.  It is saied that the anisotropic grain growth behavior by incorporating the anisotropic grain boundary energy:
![image](https://github.com/idaholab/moose/assets/61571563/fc8e94fc-f182-46de-a7fe-002764fa71e1)

However, I did not find way to define the inclination angle in GBAnisotropy. Could anybody give some suggestions?

**Comment by Hejie-Guo** [comment]
could anybody help? can moose simulate the elongate grain growth or it did need some external coding work?

**Reply by GiudGiud** [reply]
@laagesen

**Comment by laagesen** [comment]
Hi there,

Have a look at this example, as it shows how to include inclination dependence in a grain growth model:

moose/modules/phase_field/test/tests/GBAnisotropy/test3.i

You will probably also need to look at the paper which this model is based on, Phys. Rev. B, 78(2), 024113, 2008, by N. Moelans, as well as the source code for GBAnisotropy and GBAnisotropyBase.

**Reply by Hejie-Guo** [reply]
Hello, I have made some attempts, but i still need some help indeed.

1、I have tried to read the article and the source code GBAnisotropy and GBAnisotropyBase, but it is a bit difficult for me. Some equations in GBAnisotropyBase can not be found in the article and i am confused what it means.  For example, i did not find the definition in the following source code if the inclination is used and i am confused what phi_ave means and why it is equal to "pi * n / (2.0 * _op_num)".

![image](https://github.com/idaholab/moose/assets/61571563/d5bcd254-5019-4429-8efc-846e428525db)

2、I only found the following figure in the article can be related with the equation f_sigma=1+delta*cos(4phi). It seems that the source code GBAnisotropy only define the weak inclination situation. If i want to simulate the elongate grain shape, I should change the equation f_sigma=(1+delta)^(-1)*(1+delta*abs(cos(phi))) and let the delta to be 2?
![image](https://github.com/idaholab/moose/assets/61571563/1911d7b3-3a8e-473a-91cc-6cdc2725db31)


3、I also tryied to simulate the test case "moose/modules/phase_field/test/tests/GBAnisotropy/test3.i". The grain shape is consistent with the figure in the article for delta_sigma=0.5 as following figure. But the simulation did not converge when delta_sigma is bigger e.g. delta_sigma=2. 
![image](https://github.com/idaholab/moose/assets/61571563/a181d610-41ca-4105-815f-7056a374bfd2)


#Could anybody tell me more specific details and what should i do if i want simulate the elongate grain growth behaviour?

**Reply by laagesen** [reply]
You would want to use the value of epsilon prime that the authors of the paper used in Eq. (12) from the image you pasted.

**Reply by Hejie-Guo** [reply]
Sorry, the Eq.(12) is shown as following figure. I did not find the relationship between the source code and this Eq.(12). Could you please give some more details.
![image](https://github.com/idaholab/moose/assets/61571563/cb96a746-758a-407d-9acb-6c169b907912)

**Reply by laagesen** [reply]
I meant the Equation (12) from this image you posted
![image](https://github.com/idaholab/moose/assets/3810555/8168f204-e383-414a-987c-cfd55c405848)

**Reply by Hejie-Guo** [reply]
Sorry, did you mean i should set the value of delta_sigma equal to epsilon prime ? but the equation seems to be different， one use the equation f_sigma=1+delta*cons(4phi), another use the equation sigma=sigma0*(1+epsilon_prime*(cos(phi)^4+sin(phi)^4))

**Reply by laagesen** [reply]
I think you can show analytically using trigonometric relations that these two angle-dependent forms are equal to each other. Or you can just plot the functions using a polar plot to see that they are the same.

**Reply by Hejie-Guo** [reply]
Did you mean cos(4phi) equal to (cos(phi))^4+(sin(phi))^4? 
The cos(4phi) can be splitted into 2(cos(phi))^4+2(sin(phi))^4-4(cos(phi))^2*(sin(phi))^2-1， but it is not equal to  (cos(phi))^4+(sin(phi))^4.

**Reply by Hejie-Guo** [reply]
Another question is that  does this angle GBAnisotropyBase is the same with that angle in the reference? I did not understand the definition angle in GBAnisotropyBase that "phi_ave=pi*n/(2*_op_num)" and then use some other equations. could the explantation of these equations  be found in any reference?

**Reply by Hejie-Guo** [reply]
I try to use the GBanisotropiy and change the f_delta，but it seems that this parameter did not affect the results. Andthe result for the vorononi structure seem to be a homogenious state. Could anybody tell me how to model the elongate grain growth phenomenon..

here is my input file and result:

[Mesh]
  type = GeneratedMesh
  dim = 2 # Problem dimension
  nx = 12 # Number of elements in the x-direction
  ny = 12 # Number of elements in the y-direction
  nz = 0 # Number of elements in the z-direction
  xmin = 0    # minimum x-coordinate of the mesh
  xmax = 1000 # maximum x-coordinate of the mesh
  ymin = 0    # minimum y-coordinate of the mesh
  ymax = 1000 # maximum y-coordinate of the mesh
  zmin = 0
  zmax = 0
  elem_type = QUAD4 # Type of elements used in the mesh
  uniform_refine = 3 # Initial uniform refinement of the mesh

  parallel_type = replicated # Periodic BCs
[]

[GlobalParams]
  op_num = 8 # Number of grains
  var_name_base = gr # Base name of grains
[]

[UserObjects]
  [./voronoi]
    type = PolycrystalVoronoi
    grain_num = 50
    rand_seed = 100
    coloring_algorithm = bt # We must use bt to force the UserObject to assign one grain to each op
  [../]
[]

[ICs]
  [./PolycrystalICs]
    [./PolycrystalColoringIC]
      polycrystal_ic_uo = voronoi
    [../]
  [../]
[]

[Variables]
  [./PolycrystalVariables]
    # Custom action that created all of the grain variables
    order = FIRST # element type used by each grain variable
    family = LAGRANGE
  [../]
[]

[AuxVariables]
  [./bnds]
    order = FIRST
    family = LAGRANGE
  [../]
  [unique_grains]
    order = CONSTANT
    family = MONOMIAL
  []
  [var_indices]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [unique_grains]
    type = FeatureFloodCountAux
    variable = unique_grains
    flood_counter = grain_tracker
    field_display = UNIQUE_REGION
    execute_on = 'initial timestep_end'
  []
  [var_indices]
    type = FeatureFloodCountAux
    variable = var_indices
    flood_counter = grain_tracker
    field_display = VARIABLE_COLORING
    execute_on = 'initial timestep_end'
  []
[]



[Kernels]
  [./PolycrystalKernel]
  [../]
[]

[AuxKernels]
  [./bnds_aux]
    type = BndsCalcAux
    variable = bnds
    execute_on = timestep_end
  [../]
[]

[BCs]
  [./Periodic]
    [./top_bottom]
      auto_direction = 'x y' # Makes problem periodic in the x and y directions
    [../]
  [../]
[]

[Materials]
  [CuGrGr]
    type = GBAnisotropy
    Anisotropic_GB_file_name = anisotropy.txt
    inclination_anisotropy = true
    delta_sigma = 0.05
    delta_mob = 0.0
     # Quantitative material properties for copper grain growth.  Dimensions are nm and ns
    #GBmob0 = 8.5e-6 # Mobility prefactor for Cu from Schonfelder1997
    #GBenergy = 0.708 # GB energy for Cu from Schonfelder1997
    #Q = 0.53 # Activation energy for grain growth from Schonfelder 1997
    T = 1500 # Constant temperature of the simulation (for mobility calculation)
    wGB = 24 # Width of the diffuse GB
  []
[]


[UserObjects]
  [grain_tracker]
    type = GrainTracker
    tracking_step = 20 #Tracking is delayed until the polycrystalline structure is established
  []
[]

[Postprocessors]
  active = 'dt '
  [./dt]
    # Outputs the current time step
    type = TimestepSize
  [../]
[]

[Executioner]
  type = Transient # Type of executioner, here it is transient with an adaptive time step
  scheme = bdf2 # Type of time integration (2nd order backward euler), defaults to 1st order backward euler

  #Preconditioned JFNK (default)
  solve_type = 'PJFNK'

  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -mat_mffd_type'
  petsc_options_value = 'hypre    boomeramg      101                ds'

  l_max_its = 30 # Max number of linear iterations
  l_tol = 1e-4 # Relative tolerance for linear solves
  nl_max_its = 40 # Max number of nonlinear iterations
  nl_abs_tol = 1e-11 # Relative tolerance for nonlienar solves
  nl_rel_tol = 1e-8 # Absolute tolerance for nonlienar solves

  start_time = 0.0
  end_time = 4000

  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 10 # Initial time step.  In this simulation it changes.
    optimal_iterations = 6 #Time step will adapt to maintain this number of nonlinear iterations
  [../]

  [./Adaptivity]
    initial_adaptivity = 2 # Number of times mesh is adapted to initial condition
    refine_fraction = 0.7 # Fraction of high error that will be refined
    coarsen_fraction = 0.1 # Fraction of low error that will coarsened
    max_h_level = 4 # Max number of refinements used, starting from initial mesh (before uniform refinement)
  [../]
[]

[Outputs]
  exodus = true
  csv = true
  [./console]
    type = Console
    max_rows = 20
  [../]
[]

the anisotropic file:
">>>>>>>>> GB properties <<<<<<<<<
---GB energy, mobility prefactor, activation energy---
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01 6.000000e-01
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06 2.500000e-06
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01 5.300000e-01
"
****

The initial vorononi grain state:
![捕获1](https://github.com/idaholab/moose/assets/61571563/cb272140-dfaa-4c64-bae8-c3be6ff72b39)

the grain growth final state:
![捕获2](https://github.com/idaholab/moose/assets/61571563/eca1887b-a254-4d25-92c2-82362325bcbf)

**Reply by yixishen** [reply]
I believe can use GBAnisotropy function in MOOSE.  In the reference paper, inclination is coupled through grain boundary energy term, which is same as Moelans(2008). And I believe the GB orientation need to be calculated via the gradient of phase field variable. GBAnisotropy in MOOSE is based on Phys. Rev. B, 78(2), 024113, 2008, by N. Moelans, which is ref 33 in the reference paper you mentioned. The methods to calculate PF parameter from GB energy is same between your reference and Moelans work.
